smolyay/smolyak.py

```diff
@@ -0,0 +1,158 @@
+import numpy as np
+import scipy.special
+
+
+class IndexGrid:
+    """Generate index of grid points.
+
+    Attributes
+    ----------
+    dimension : int
+        Number of independent variables.
+    exactness : int
+        Level of exactness of the approximation.
+    index_per_level : list
+        Number of indexes at each level
+    """
+
+    def __init__(self, dimension, exactness, index_per_level):
+        self.dimension = dimension
+        self.exactness = exactness
+        self.index_per_level = index_per_level
+
+        if (type(index_per_level) == tuple or
+                type(index_per_level) == np.ndarray):
+            index_per_level = list(index_per_level)
+
+        if len(index_per_level) != exactness + 1:
+            raise IndexError(
+                "index_per_level must be an array with a length of {}"
+                .format(exactness + 1))
+
+    def level_indexes(self):
+        """Generate indexes of levels.
+
+        Returns
+        -------
+        level_indexes : numpy array
+            Unique indexes at each level.
+
+        """
+        # creating an array to store unique indexes in each level
+        level_indexes = np.zeros((self.exactness+1), dtype=list)
+
+        max_index = sum(self.index_per_level)
+        indexes = np.linspace(0, max_index, max_index+1, dtype=np.int32)
+
+        # intial level
+        level_indexes[0] = indexes[0:self.index_per_level[0]]
+
+        # storing unique indexes for each level
+        summ_index = self.index_per_level[0]
+        for idx_index_per_level in range(1, self.exactness + 1):
+            summ_index += self.index_per_level[idx_index_per_level]
+            level_indexes[idx_index_per_level] = (
+                indexes[summ_index-self.index_per_level[idx_index_per_level]:
+                        summ_index])
+
+        return level_indexes
+
+    def grid_point_index(self):
+        """Generate index of grid points.
+
+        First, it generates allowed composoitons of levels
+        via NEXCOM algorithm. Then, it replaces the levels
+        with their corresponding indexes, and
+        makes the index for the grid points.
+
+        Returns
+        -------
+        grid_points_index : numpy array
+            Index of grid points
+
+        """
+        num_grid_points = 0
+        level_indexes = self.level_indexes()
+
+        # precompute the number of level compositions
+        summs = np.linspace(self.dimension,
+                            self.exactness+self.dimension,
+                            self.exactness+1, dtype=np.int32)
+        max_level_indexes = summs - self.dimension
+        num_comps = 0
+        for max_level_index in max_level_indexes:
+            num_comps += scipy.special.comb(max_level_index+self.dimension-1,
+                                            self.dimension-1, exact=True)
+        level_compositions = np.zeros((num_comps, self.dimension), dtype=list)
+
```

Comment:
For the possible level_compositions, is there ever going to be a reason that we want the code to print that out so it can be seen for writing a paper or any other reason? If so, this function may need to be split into multiple functions.

Reply:
For the case of generating grid points, I don't think it's necessary; however, we can split it in case it's needed.